### PR TITLE
Fix undefined behavior in MuonIdProducer from empty pointer

### PR DESCRIPTION
**Pull request by trocino**
#### PR description:

Fix undefined behavior in `MuonIdProducer`, reported in issue [35036](https://github.com/cms-sw/cmssw/issues/35036). This is caused by the `theShowerDigiFiller_` pointer, which is used empty to call function `fillDefault(...)` whenever the `fillShowerDigis_` flag is false, e.g. in HLT.

I added a default constructor without parameters for `MuonShowerDigiFiller`, which is only used when `fillShowerDigis_` is false. Function `fillDefault(...)` doesn't use any data member of `MuonShowerDigiFiller` and can be called even with this minimal initialization.
In one [point](https://github.com/trocino/cmssw/blob/fix_UBSAN_issue35036_MUO-MuIdProd/RecoMuon/MuonIdentification/plugins/MuonIdProducer.cc#L894) I also had to fix the condition to use the fully initialized `MuonShowerDigiFiller`.

#### PR validation:

The code compiles without errors and passes the basic tests in `CMSSW_12_1_X_2021-09-15-1100`. I verified that the undefined behavior error disappears, running wf 11603.0 in `CMSSW_12_1_UBSAN_X_2021-09-13-1100`.

### CODE REVIEW

File: RecoMuon/MuonIdentification/interface/MuonShowerDigiFiller.h [hunk 1]
```diff
@@ -46,6 +46,7 @@
 
 class MuonShowerDigiFiller {
 public:
+  MuonShowerDigiFiller() {}
```
Comment (slava77) [inline RecoMuon/MuonIdentification/interface/MuonShowerDigiFiller.h:49]: is this really needed with an empty body? it should already be the same as the implicit default.
please drop.

Comment (trocino) [reply RecoMuon/MuonIdentification/interface/MuonShowerDigiFiller.h:49]: Is there an implicit _default_ constructor? If I remove this line, I get a compilation error at the instantiation without parameters [here](https://github.com/trocino/cmssw/blob/fix_UBSAN_issue35036_MUO-MuIdProd/RecoMuon/MuonIdentification/plugins/MuonIdProducer.cc#L93). From the error message, it looks like there is an implicit copy constructor instead:

`/.../unique_ptr.h: In instantiation of 'typename std::_MakeUniq<_Tp>::__single_object std::make_unique(_Args&& ...) [with _Tp = MuonShowerDigiFiller; _Args = {}; typename std::_MakeUniq<_Tp>::__single_object = std::unique_ptr<MuonShowerDigiFiller>]':`

`/.../plugins/MuonIdProducer.cc:93:67:   required from here`
`/.../unique_ptr.h:962:30: error: no matching function for call to 'MuonShowerDigiFiller::MuonShowerDigiFiller()'`
`  962 |     { return unique_ptr<_Tp>(new _Tp(std::forward<_Args>(__args)...)); }`

`In file included from /.../plugins/MuonIdProducer.h:49,`
`                 from /.../plugins/MuonIdProducer.cc:12:`
`/.../interface/MuonShowerDigiFiller.h:49:3: note: candidate: 'MuonShowerDigiFiller::MuonShowerDigiFiller(const edm::ParameterSet&, edm::ConsumesCollector&&)'`
`   49 |   MuonShowerDigiFiller(const edm::ParameterSet&, edm::ConsumesCollector&& iC);`
`/.../interface/MuonShowerDigiFiller.h:49:3: note:   candidate expects 2 arguments, 0 provided`

`/.../interface/MuonShowerDigiFiller.h:47:7: note: candidate: 'MuonShowerDigiFiller::MuonShowerDigiFiller(const MuonShowerDigiFiller&)'`
`   47 | class MuonShowerDigiFiller {`
`/.../interface/MuonShowerDigiFiller.h:47:7: note:   candidate expects 1 argument, 0 provided`

`/.../interface/MuonShowerDigiFiller.h:47:7: note: candidate:` `'MuonShowerDigiFiller::MuonShowerDigiFiller(MuonShowerDigiFiller&&)'`
`/.../interface/MuonShowerDigiFiller.h:47:7: note:   candidate expects 1 argument, 0 provided`

Comment (slava77) [inline RecoMuon/MuonIdentification/interface/MuonShowerDigiFiller.h:49]: thanks for checking.
I forgot that the implicit default is created only if there is no other kind defined.

**Comment by slava77** [comment]
@cmsbuild please test

**Comment by slava77** [comment]
+reconstruction

for https://github.com/cms-sw/cmssw/pull/35303	2577c56 
- code changes are in line with the PR description and should address the UBSAN issue based on just the visual inspection (even though it's not obvious why the code calling `fillDefault` was not crashing already )
- jenkins tests pass and comparisons with the baseline show no (relevant) differences

**Comment by perrotta** [comment]
+1

**Pull request merged**